doc: genrest: Use Kconfiglib 2 and improve output

ulfalizer opened the pull request:
The new version of the reference page for a symbol lists the following
essential and generally useful information at the top of the page:

 - Prompt

 - Type

 - Help text

 - Direct dependencies (from 'depends on', and inherited from if's and
   menus)

 - Defaults

 - Other symbols selecting or implying the symbol

The current symbol value and visibility are no longer displayed. They
don't make much sense in a reference document, as they depend on user
values from configuration files.

The bottom of the reference page also includes the symbol represented in
Kconfig format (including propagated dependencies).

All references to defined symbols appearing in expressions are turned
into links, which can be clicked on to jump to the reference page for
that symbol.

The following (slightly outdated) screenshot gives an idea of the
format:
https://www.dropbox.com/s/a34tlk2ncyus8po/promptandtype.png?dl=0

This commit also removes Kconfiglib 1, since it is no longer used.

Fixes #5622
Fixes #6821

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

ulfalizer committed:
doc: genrest: Use Kconfiglib 2 and improve output

The new version of the reference page for a symbol lists the following
essential and generally useful information at the top of the page:

 - Prompt

 - Type

 - Help text

 - Direct dependencies (from 'depends on', and inherited from if's and
   menus)

 - Defaults

 - Other symbols selecting or implying the symbol

The current symbol value and visibility are no longer displayed. They
don't make much sense in a reference document, as they depend on user
values from configuration files.

The bottom of the reference page also includes the symbol represented in
Kconfig format (including propagated dependencies).

All references to defined symbols appearing in expressions are turned
into links, which can be clicked on to jump to the reference page for
that symbol.

The following (slightly outdated) screenshot gives an idea of the
format:
https://www.dropbox.com/s/a34tlk2ncyus8po/promptandtype.png?dl=0

This commit also removes Kconfiglib 1, since it is no longer used.

Fixes #5622
Fixes #6821

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

mbolivar commented:
> All references to defined symbols appearing in expressions are turned
> into links, which can be clicked on to jump to the reference page for
> that symbol.

Does this include symbols selected by a given config?

ulfalizer commented:
@mbolivar 
I was just going to create a separate issue for that. It is broken at the moment, but will be fixed once https://github.com/ulfalizer/Kconfiglib/commit/a5ea0dcbc8b9343f6901654ad92de6ece156f954 gets in.

I also want to have `A && B || C && D` rendered as `(A && B) || (C && D)` for clarity (bit silly to require people to memorize the precedence table). That'll be rolled into the next Kconfiglib update as well.

ulfalizer commented:
@mbolivar https://github.com/zephyrproject-rtos/zephyr/issues/6844

dbkinder commented on doc/scripts/genrest/genrest.py:
```diff
@@ -96,15 +63,148 @@ def print_items(items, outdir, indent):
+                # and the like
+                sym_rst.write("Help\n"
+                              "====\n\n"
+                              ".. code-block:: none\n\n"
```
I think there are folks using markup tags in the help text for lists and such that won't render as expected if treated as a code block.

ulfalizer replied:
@dbkinder 
Not sure about Zephyr, but I've seen quite a lot of help texts with figures drawn in text in the Linux kernel.

I agree that the output is uglier compared to a plain paragraph.

ulfalizer replied:
Oh, you mean they use RST formatting in help texts? Yeah, in that case, it's definitely wrong to wrap it in a `code-block`.

ulfalizer replied:
Talked a bit to @mbolivar on IRC. It wouldn't take anything fancy for the formatting to get messed up either. Take a look at `SYS_LOG_SPI_LEVEL` in `drivers/spi/Kconfig` for example.

Looks like you can preserve newlines in RST by putting `|` at the beginning of each line. There are probably other cases that would get messed up still though. I think a `code-block` would be safest.

nashif replied:
I think the help text should be presented as text like it used to be, using code-block here looks really bad. The few cases were this get messed up we can live with or fix in Kconfig.

ulfalizer commented:
Another slightly wonky thing that's worth mentioning:

If a symbol is defined in multiple locations, all the properties that are tied to the symbol itself (the `default`s, `select`s, `imply`s, and `range`s), as opposed to being tied to the menu node (e.g. `prompt` and `help`), will be displayed on the first symbol definition in the Kconfig display. It's tricky to work around, as that's how Kconfig sees things internally.

I wonder if that's confusing enough to be worth mentioning on the reference pages.

ulfalizer commented on doc/scripts/genrest/genrest.py:
```diff
@@ -84,9 +48,12 @@ def print_items(items, outdir, indent):
 :program:`Kconfig` using the :file:`~/doc/scripts/genrest/genrest.py` script.
-Click on the option name in the table below for detailed information about
-each option.
+Click on the symbol name in the table below for to see its definition.
```
Heh... not sure why I changed this. Leftover from some earlier rewriting I think. I could change it back.

ulfalizer commented:
> Another slightly wonky thing that's worth mentioning:
> 
> If a symbol is defined in multiple locations, all the properties that are tied to the symbol itself (the defaults, selects, implys, and ranges), as opposed to being tied to the menu node (e.g. prompt and help), will be displayed on the first symbol definition in the Kconfig display. It's tricky to work around, as that's how Kconfig sees things internally.
> 
> I wonder if that's confusing enough to be worth mentioning on the reference pages.

Just remembered that I mentioned it in a new paragraph in `index.rst`.

dbkinder commented:
We’re not using a code block today and haven’t noticed a problem.

-- david kinder

On Mar 28, 2018, at 12:48 PM, Ulf Magnusson <notifications@github.com<mailto:notifications@github.com>> wrote:


@ulfalizer commented on this pull request.

________________________________

In doc/scripts/genrest/genrest.py<https://github.com/zephyrproject-rtos/zephyr/pull/6843#discussion_r177868980>:

> +                      ".. title:: {0}\n\n"
+                      ".. option:: CONFIG_{0}\n\n"
+                      "{1}\n\n"
+                      "Type: ``{2}``\n\n"
+                      .format(sym.name,
+                              prompt_str,
+                              kconfiglib.TYPE_TO_STR[sym.type]))
+
+        # Symbols with multiple definitions can have multiple help texts
+        for node in sym.nodes:
+            if node.help is not None:
+                # Wrap the help text in a code block to preserve "ASCII art"
+                # and the like
+                sym_rst.write("Help\n"
+                              "====\n\n"
+                              ".. code-block:: none\n\n"


Talked a bit to @mbolivar<https://github.com/mbolivar> on IRC. It wouldn't take anything fancy for the formatting to get messed up either. Take a look at SYS_LOG_SPI_LEVEL in drivers/spi/Kconfig for example.

Looks like you can preserve newlines in RST by putting | at the beginning of each line. There are probably other cases that would get messed up still though. I think a code-block would be safest.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/zephyrproject-rtos/zephyr/pull/6843#discussion_r177868980>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKjRQue5p9NzO5P7Af5EQtkmMXaGtVTOks5ti-j_gaJpZM4S_BbW>.

ulfalizer commented:
@dbkinder 
Maybe it wasn't obviously broken if you didn't know what it was supposed to look like.

Here's how the old version of `genrest.py` renders the help text for `SYS_LOG_SPI_LEVEL` for example:
https://www.dropbox.com/s/kojxv0netilyaau/brokenhelp.png?dl=0 (all mangled up into a single paragraph)

Here's what it's supposed to look like (`drivers/spi/Kconfig`):

```
Sets log level for SPI drivers.
Levels are:
0 OFF, do not write
1 ERROR, only write SYS_LOG_ERR
2 WARNING, write SYS_LOG_WRN in addition to previous level
3 INFO, write SYS_LOG_INF in addition to previous levels
4 DEBUG, write SYS_LOG_DBG in addition to previous levels
```

Seems you can put a `|` at the beginning of each line in RST to force line breaks, which would fix that particular case, but that would still break stuff like alignment and (maybe) indentation.

mbolivar reviewed (approved):
I built this, looked at the results, and I think it's a solid improvement.

There are ugly scrollbars in the kconfig definitions for some options, but I think that's tolerable and can be fixed later. The benefit of the more appropriate data and otherwise improved formatting seems worth having.

dbkinder commented:
Ah.  I see. I was trying to look at this with my phone while at the airport.  You can see the currently rendered output at http://docs.zephyrproject.org/reference/kconfig/CONFIG_SYS_LOG_SPI_LEVEL.html#cmdoption-arg-config-sys-log-spi-level

Yep.  The output is indeed just flattened out into one paragraph.  Your change to display the help as a fixed-width font code block seems overkill though.  My preference would be to keep the format as is today and then spend the time to edit the Kconfig files to improve how they're written so their output as HTML looks better.  (As stated above, fit it later.)

carlescufi reviewed (approved):
+1 to merging this

nashif commented:
![image](https://user-images.githubusercontent.com/180017/38087710-26d91c22-3327-11e8-9b40-f28887909d79.png)

with 
```
config SYS_LOG_SPI_LEVEL
        int
        prompt "SPI Driver Log level"
        depends on SYS_LOG && SPI
        default 0
        range 0 4
        help
          Sets log level for SPI drivers.  Levels are:

            0 OFF, do not write

            1 ERROR, only write SYS_LOG_ERR

            2 WARNING, write SYS_LOG_WRN in addition to previous level

            3 INFO, write SYS_LOG_INF in addition to previous levels

            4 DEBUG, write SYS_LOG_DBG in addition to previous levels

```

ulfalizer committed:
drivers: spi: Format SYS_LOG_SPI_LEVEL help nicer

The previous help text rendered as a single paragraph in the Kconfig RST
reference. The new help text renders as a bullet list and matches the
format of SYS_LOG_CRYPTO_LEVEL.

Signed-off-by: Ulf Magnusson <ulfalizer@gmail.com>

ulfalizer commented:
I removed the `code-block`.

Piggybacking a formatting fix for `SYS_LOG_SPI_LEVEL` as well, which makes it consistent with `SYS_LOG_CRYPTO_LEVEL`. It now renders like this:

![spihelp](https://user-images.githubusercontent.com/154768/38099044-e3c04a98-3379-11e8-8bfd-34b375edfbf3.png)

dbkinder commented:
Thanks!

-- david kinder

On Mar 29, 2018, at 8:52 AM, Ulf Magnusson <notifications@github.com<mailto:notifications@github.com>> wrote:


I removed the code-block.

Piggybacking a formatting fix for SYS_LOG_SPI_LEVEL as well, which makes it consistent with SYS_LOG_CRYPTO_LEVEL. It now renders like this:

[spihelp]<https://user-images.githubusercontent.com/154768/38099044-e3c04a98-3379-11e8-8bfd-34b375edfbf3.png>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/zephyrproject-rtos/zephyr/pull/6843#issuecomment-377281278>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AKjRQl3Qqzrd1J4Fyip5eMTTQN1s-afSks5tjQNCgaJpZM4S_BbW>.
